ipa-core/src/report.rs

```diff
@@ -413,6 +434,186 @@ where
     }
 }
 
+/// A binary report as submitted by a report collector, containing encrypted `OprfReport`
+/// An `EncryptedOprfReport` consists of:
+///     `ct_mk`: Enc(`match_key`)
+///     `ct_btt`: Enc(`breakdown_key`, `trigger_value`, `timestamp`)
+///     associated data of `ct_mk`: `key_id`, `epoch`, `event_type`, `site_domain`,
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct EncryptedOprfReport<BK, TV, TS, B>
+where
+    B: Deref<Target = [u8]>,
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue,
+{
+    data: B,
+    phantom_data: PhantomData<(BK, TV, TS)>,
+}
+
+/// follows the outline of the implementation of `EncryptedReport`
+impl<B, BK, TV, TS> EncryptedOprfReport<BK, TV, TS, B>
+where
+    B: Deref<Target = [u8]>,
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue,
+    Replicated<BK>: Serializable,
+    Replicated<TV>: Serializable,
+    Replicated<TS>: Serializable,
+{
+    const ENCAP_KEY_MK_OFFSET: usize = 0;
+    const CIPHERTEXT_MK_OFFSET: usize = Self::ENCAP_KEY_MK_OFFSET + EncapsulationSize::USIZE;
+    // need to round up ciphertext length to nearest multiple of 16 since AES has block length 16
```

Comment:
Our current AEAD is AES-GCM, which does not expand the ciphertext to a multiple of the block size, but coincidentally `Replicated<BA64>` uses exactly one block. To work with other sizes something like `BTT_END` would be needed in the encryption routine.

If we do have padding, my preference would be to reject inputs where the padding is not all zero. Given that we want to make some of these values runtime-configurable in the future, padding does give us a bit of flexibility to tweak the format, although my preference would be to have a version indicator at the very start of the report.

Reply:
I agree. I think the size of the ciphertext is simply ENC_KEY + PAYLOAD + AAD_TAG. I would prefer to get the tag size and key size from HPKE module rather than hardcoding them here

Reply:
AES-GCM uses a counter mode, it also uses a 128 bit multiplication over a Galois-Field to compute the tag (see https://en.wikipedia.org/wiki/Galois/Counter_Mode). I agree that the ciphertext wouldn't need to be a multiple of 128 bit block.  However, for the ciphertext encrypting trigger value, breakdown, ... the plaintext is not coincidentally a multiple of 128 bit blocks, nevertheless its ciphertext is (or at least seems to be). Therefore the ciphertext size seems to be ENC_KEY + PAYLOAD rounded up to a multiple of 128 bits + AAD_TAG. Previously, I did not round it up and it caused a panic in the assertion debug_assert_eq!(out.len(), usize::from(self.encrypted_len())); in the encrypt functionality of OprfReport.

Reply:
Is the ciphertext coming out a multiple of 16 bytes not explained by the size of the plaintext we're supplying?

```
let mut plaintext_btt = vec![0u8; Self::BTT_END];
```

I think the following two changes would make the code work for any size match key up to 64 bits (after updating the necessary references to `BA64`):

Line 636:
```
            match_key: Replicated::<BA64>::deserialize(GenericArray::from_slice(
                &plaintext_mk[0..(<Replicated<BA64> as Serializable>::Size::USIZE)],
            ))
```

Line 825:
```
        let mut plaintext_mk = vec![0u8; 16];
        self.match_key.serialize(GenericArray::from_mut_slice(
            &mut plaintext_mk[0..(<Replicated<BA64> as Serializable>::Size::USIZE)],
        ));
```

Reply:
I dont think so but there might be something odd going on. Right now, in 
```
#[test]
    fn enc_dec_roundtrip_oprf()
```
I use 
```
let report = OprfReport::<BA8, BA3, BA20>
```
and compute
```
let enc_report_bytes = report.encrypt(key_id, &key_registry, &mut rng).unwrap();
let enc_report = EncryptedOprfReport::from_bytes(enc_report_bytes.as_slice()).unwrap();
let dec_report: OprfReport<BA8, BA3, BA20> = enc_report.decrypt(&key_registry).unwrap();
```
I get an error message that the ciphertext length doesn't match when I don't round up $2\cdot (8+3+20)=62$ bit to  $16$ byte

Reply:
I dont think so, in 
```
    #[test]
    fn test_decryption_fails() 
```
I get and error if I dont round up to a multiple of 16 even though the plaintext contains 62 bit:
```
let report = OprfReport::<BA8, BA3, BA20> 
let enc_report_bytes = report.encrypt(key_id, &key_registry, &mut rng).unwrap();
let enc_report = EncryptedOprfReport::from_bytes(enc_report_bytes.as_slice()).unwrap();
let dec_report: OprfReport<BA8, BA3, BA20> = enc_report.decrypt(&key_registry).unwrap();

assert_eq!(dec_report, report);
```